### PR TITLE
Truffle: reduce footprint of node rewriting assumption

### PR DESCRIPTION
**Pull request by woess**
The goals of this pull request are:
1. avoid the string concatenation for the assumption name `"nodeRewritingAssumption of " + rootNode`, unless the `TraceTruffleAssumptions` option is enabled (assuming it's still useful).
2. inline the `CyclicAssumption` object into `OptimizedCallTarget`
3. allocate the node rewriting assumption lazily (when actually needed by the compiler).

### CODE REVIEW

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java [hunk 1]
```diff
@@ -147,7 +151,34 @@ private static RootNode cloneRootNode(RootNode root) {
...
+            // if CAS failed, assumption is already initialized; cannot be null after that.
+            return Objects.requireNonNull(nodeRewritingAssumption);
+        }
+    }
+
+    /**
+     * Invalidate node rewriting assumption iff it has been initialized.
+     */
+    private void invalidateNodeRewritingAssumption() {
+        Assumption oldAssumption = NODE_REWRITING_ASSUMPTION_UPDATER.getAndUpdate(this, prev -> prev == null ? null : runtime.createAssumption(prev.getName()));
```
Comment (chumer) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: Please no lambdas. Footprint....

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java [hunk 2]
```diff
@@ -94,7 +97,7 @@
     @CompilationFinal private Assumption profiledReturnTypeAssumption;
 
     private final RootNode rootNode;
-    private volatile RootNode uninitializedRootNode = UNINITIALIZED;
+    private RootNode uninitializedRootNode = UNINITIALIZED;
```
Comment (chumer) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: Why that? Separate PR?

**Comment by chumer** [comment]
Besides my complaints. Looks good.

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java (hunk 2 above)
Comment (woess) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: I'll remove it

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java (hunk 1 above)
Comment (woess) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: ok, I'll replace it with an anonymous class

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java [hunk 3]
```diff
@@ -106,7 +110,9 @@
      * assumption. It gets invalidated when a node rewriting is performed. This ensures that all
      * compiled methods that have this call target inlined are properly invalidated.
      */
-    private final CyclicAssumption nodeRewritingAssumption;
+    private volatile Assumption nodeRewritingAssumption;
+    private static final AtomicReferenceFieldUpdater<OptimizedCallTarget, Assumption> NODE_REWRITING_ASSUMPTION_UPDATER =
+                    AtomicReferenceFieldUpdater.newUpdater(OptimizedCallTarget.class, Assumption.class, "nodeRewritingAssumption");
```
Comment (jtulach) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: Is it a generic problem that `CyclicAssumption` provides *yet another level of indication* and is it interesting to solve it? Btw. I had an experiment cleaning up assumptions somewhere.

**Comment by chumer** [comment]
Now that I think again. It could use some tests. Especially around thread-safety.

File: graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java (hunk 3 above)
Comment (woess) [inline graal/com.oracle.graal.truffle/src/com/oracle/graal/truffle/OptimizedCallTarget.java]: `OptimizedCallTarget` is such a central and heavily used part of Truffle that I think it deserves a bit of microoptimization. I don't think there's a high number of `CyclicAssumption` instances in the current language implementations, so I don't think it's a general problem.

**Comment by woess** [comment]
@chumer I've added a test, please review.

File: graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            return (int) callNode.call(frame, frame.getArguments());
+        }
+    }
+
+    @Test
+    public void testRewriteAssumption() {
+        String testName = "testRewriteAssumption";
+        final int compilationThreshold = TruffleCompilerOptions.TruffleCompilationThreshold.getValue();
+        assumeTrue(compilationThreshold >= 2);
+        assumeTrue(TruffleCompilerOptions.TruffleFunctionInlining.getValue());
```
Comment (chumer) [inline graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java]: I would assert this otherwise it will just not test it silently if these options change.

File: graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import com.oracle.truffle.api.nodes.DirectCallNode;
+
+public class OptimizedCallTargetTest {
+    private static final GraalTruffleRuntime runtime = (GraalTruffleRuntime) Truffle.getRuntime();
+    private static final Field nodeRewritingAssumptionField;
+    static {
+        try {
+            nodeRewritingAssumptionField = OptimizedCallTarget.class.getDeclaredField("nodeRewritingAssumption");
+            nodeRewritingAssumptionField.setAccessible(true);
+        } catch (NoSuchFieldException | SecurityException e) {
```
Comment (chumer) [inline graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java]: I'd prefer renaming the package name instead and use package-protected API. But this should be a separate PR.

File: graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java (hunk 4 above)
Comment (woess) [inline graal/com.oracle.graal.truffle.test/src/com/oracle/graal/truffle/test/OptimizedCallTargetTest.java]: ok, I'll assert.
